### PR TITLE
Added attribute modules for o365 sendAs/onBehalf

### PR DESCRIPTION
**Pull request by metodej**
- added modules for storing ids of subgroups that check whether there are indeed ids of subgroups stored
- added virtual modules that calculates whether member has right to send as or on behalf of groups
- also added tests for the check and the calculation

### CODE REVIEW

**Review by Johaney-s — CHANGES_REQUESTED** [review]
Really nice piece of code, only small suggestions from me :)

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_group_attribute_def_def_o365SendOnBehalfGroups.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.ArrayList;
+
+/**
+ * @author Metodej Klang <metodej.klang@gmail.com>
+ */
+public class urn_perun_group_attribute_def_def_o365SendOnBehalfGroups extends GroupAttributesModuleAbstract implements GroupAttributesModuleImplApi {
+
+	@Override
+	public void checkAttributeSyntax(PerunSessionImpl sess, Group group, Attribute attribute) throws WrongAttributeValueException {
+		for (String groupId : attribute.valueAsList()) {
```
[the remark this answers is not in the document]
Comment (metodej) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_group_attribute_def_def_o365SendOnBehalfGroups.java:26]: Fixed.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_group_attribute_def_def_o365SendAsGroups.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.ArrayList;
+
+/**
+ * @author Metodej Klang <metodej.klang@gmail.com>
+ */
+public class urn_perun_group_attribute_def_def_o365SendAsGroups extends GroupAttributesModuleAbstract implements GroupAttributesModuleImplApi {
+
+	@Override
+	public void checkAttributeSyntax(PerunSessionImpl sess, Group group, Attribute attribute) throws WrongAttributeValueException {
+		for (String groupId : attribute.valueAsList()) {
```
[the remark this answers is not in the document]
Comment (metodej) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_group_attribute_def_def_o365SendAsGroups.java:26]: Fixed.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ModulesUtilsBlImpl.java [hunk 3]
```diff
@@ -1198,6 +1200,49 @@ public User getUserByLoginInNamespace(PerunSession sess, String login, String na
...
+			if (sendAs.getValue() != null && sendAs.valueAsBoolean()) {
+				return true;
+			}
+
+			Attribute sendAsGroups = sess.getPerunBl().getAttributesManagerBl().getAttribute(sess, group, listAttribute);
+			List<String> subgroups = sendAsGroups.valueAsList();
+
+			for (String groupId : subgroups) {
+				Group subgroup = sess.getPerunBl().getGroupsManagerBl().getGroupById(sess, Integer.parseInt(groupId));
+				if (sess.getPerunBl().getGroupsManagerBl().getGroupMembers(sess, subgroup).contains(member)) {
```
[the remark this answers is not in the document]
Comment (metodej) [reply perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ModulesUtilsBlImpl.java]: Fixed.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ModulesUtilsBlImpl.java [hunk 4]
```diff
@@ -1198,6 +1200,49 @@ public User getUserByLoginInNamespace(PerunSession sess, String login, String na
 		}
 	}
 
+	@Override
+	public boolean getSendRightFromAttributes(PerunSessionImpl sess, Member member, Group group, String booleanAttribute, String listAttribute) {
+		try {
+			Attribute sendAs = sess.getPerunBl().getAttributesManagerBl().getAttribute(sess, member, group, booleanAttribute);
+			if (sendAs.getValue() != null && sendAs.valueAsBoolean()) {
```
[the remark this answers is not in the document]
Comment (metodej) [reply perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ModulesUtilsBlImpl.java:1207]: Added check.

**Review by Johaney-s — APPROVED** [review]

**Review by cuadradek — COMMENTED** [review]
I think we should add tests also for each attribute module. What do you think @metodej?

**Comment by metodej** [comment]
> 
> 
> I think we should add tests also for each attribute module. What do you think @metodej?

I did not think it necessary since the shared methods are tested in `ModulesUtilsEntryIntegrationTest` and the modules pretty much just call these methods. But I guess I could add a test of that syntax check for the two modules,,,

**Review by cuadradek — APPROVED** [review]

**Pull request merged**